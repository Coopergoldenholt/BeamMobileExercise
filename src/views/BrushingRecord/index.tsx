import AsyncStorage from '@react-native-async-storage/async-storage'
import React, { useEffect, useState } from 'react'
import { Alert, StyleSheet, Text, TextInput, TouchableOpacity, View } from 'react-native'
import DatePicker from 'react-native-date-picker'
import { formatDateTime } from '../../functions/formatDateTime'
import { IBrushingRecord } from '../../types'
import uuid from 'react-native-uuid';

interface BrushingRecordPros {
    navigation: any;
    route: any;
}

const BrushingRecord: React.FC<BrushingRecordPros> = ({ navigation, route, ...props }) => {
    const [displayTimePicker, setDisplayTimePicker] = useState(false)
    const [date, setDate] = useState(new Date())
    const [duration, setDuration] = useState({ minutes: '0', seconds: '0' })

    useEffect(() => {
        if (route.params) {
            const { date, duration } = route.params.record
            setDate(new Date(date))
            setDuration(duration)
        }
    }, [route])

    const storeBrushingRecord = async () => {
        const brushingRecord: IBrushingRecord = {
            //@ts-ignore
            id: uuid.v4(),
            date,
            duration
        }
        let brushingRecords: any = []

        try {
            const value = await AsyncStorage.getItem('brushingRecords')
            if (value !== null) {
                brushingRecords = JSON.parse(value)
            }
        } catch (e) {
            Alert.alert('Could not save your brushing record.')
        }

        brushingRecords.unshift(brushingRecord)

        try {
            const stringifiedRecords = JSON.stringify(brushingRecords)
            await AsyncStorage.setItem('brushingRecords', stringifiedRecords)
            navigation.goBack()
        } catch (e) {
            Alert.alert('Could not save your brushing record.')
        }
    }

    const updateBrushingRecord = async () => {
        const brushingRecord: IBrushingRecord = {
            id: route.params.record.id,
            date,
            duration
        }

        let brushingRecords: any = []

        try {
            const value = await AsyncStorage.getItem('brushingRecords')
            if (value !== null) {

                brushingRecords = JSON.parse(value)
            }
        } catch (e) {
            Alert.alert('Could not update your brushing record.')
        }

        const updatedRecords = brushingRecords.map((rec: IBrushingRecord) => {
            if (rec.id == route.params.record.id) {
                return brushingRecord
            }
            return rec
        })

        try {
            const stringifiedRecords = JSON.stringify(updatedRecords)
            await AsyncStorage.setItem('brushingRecords', stringifiedRecords)

        } catch (e) {
            Alert.alert('Could not update your brushing record.')
        }

        navigation.goBack()
    }

    return <View style={{ padding: 15, paddingTop: 50 }}>
        <Text testID='Header.Text' style={styles.headerText}>{route.params ? 'Edit Record' : 'Add Record'}</Text>
        <DatePicker
            mode='datetime'
            modal
            open={displayTimePicker}
            date={date}
            onConfirm={(date) => {
                setDisplayTimePicker(false)
                setDate(date)
            }}
            onCancel={() => {
                setDisplayTimePicker(false)
            }}
        />
        <TouchableOpacity
            style={styles.dateDisplay}
            onPress={() => setDisplayTimePicker(!displayTimePicker)}
        >
            <Text style={styles.dateText}>{formatDateTime(date)}</Text>
        </TouchableOpacity>
        <Text style={styles.durationText}>Duration:</Text>
        <View style={styles.rowView}>
            <View style={styles.inputContainer}>
                <TextInput
                    placeholder='Minutes'
                    style={styles.durationInput}
                    value={duration.minutes}
                    keyboardType='number-pad'
                    onChangeText={(minutes) => setDuration({ ...duration, minutes })}
                />
                <Text>Minutes</Text>
            </View>
            <View style={styles.inputContainer}>
                <TextInput
                    placeholder='Seconds'
                    style={styles.durationInput}
                    value={duration.seconds}
                    keyboardType='number-pad'
                    onChangeText={(seconds) => setDuration({ ...duration, seconds })}
                />
                <Text>Seconds</Text>
            </View>
        </View>

        <View style={{ ...styles.rowView, paddingTop: 20 }}>
            <TouchableOpacity
                style={styles.cancelButton}
                onPress={() => navigation.goBack()}>
                <Text>Cancel</Text>
            </TouchableOpacity>
            <TouchableOpacity
                testID='SaveRecord.Button'
                style={styles.button}
                onPress={route.params ? () => updateBrushingRecord() : () => storeBrushingRecord()}
            >
                <Text style={styles.buttonText}>Save</Text>
            </TouchableOpacity>
        </View>
    </View>
}

const styles = StyleSheet.create({
    headerText: {
        fontSize: 22,
        alignSelf: 'center',
        paddingBottom: 25,
        fontWeight: 'bold'
    },
    rowView: {
        flexDirection: 'row',
        justifyContent: 'space-between',
    },
    dateText: {
        fontSize: 20
    },
    durationText: {
        fontSize: 18,
        paddingTop: 20,
        paddingBottom: 10
    },
    dateDisplay: {
        borderColor: 'grey',
        borderWidth: 1,
        borderRadius: 5,
        height: 50,
        justifyContent: 'center',
        paddingLeft: 5,
        backgroundColor: 'white'
    },
    inputContainer: {
        flexDirection: 'row',
        justifyContent: 'center',
        alignItems: 'center'
    },
    durationInput: {
        borderColor: 'grey',
        borderWidth: 2,
        minWidth: 75,
        height: 40,
        borderRadius: 5,
        marginRight: 5,
        paddingLeft: 5,
        backgroundColor: 'white'
    },
    button: {
        alignSelf: 'center',
        backgroundColor: '#20a9ff',
        borderRadius: 5,
        height: 40,
        width: 150,
        alignItems: 'center',
        justifyContent: 'center',
    },
    cancelButton: {
        alignSelf: 'center',
        backgroundColor: 'tomato',
        borderRadius: 5,
        height: 40,
        width: 150,
        alignItems: 'center',
        justifyContent: 'center',
    },
    buttonText: {
        color: 'white'
    },
})

export default BrushingRecord