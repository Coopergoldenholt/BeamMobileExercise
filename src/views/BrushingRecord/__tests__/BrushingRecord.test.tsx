import React from 'react'
import { fireEvent, render } from '@testing-library/react-native'
import BrushingRecord from '..'
import AsyncStorage from '@react-native-async-storage/async-storage';

const route = {
    params: {
        record: {
            date: new Date(),
            duration: { minutes: '2', seconds: '34' }
        }
    }
}

const navigation = {
    navigate: jest.fn(),
    goBack: jest.fn()
};

test('renders correctly', () => {
    const { toJSON } = render(<BrushingRecord navigation route={route} />)
    expect(toJSON()).toMatchSnapshot();
})


test('Button fires with params', async () => {

    const { getByTestId } = render(<BrushingRecord navigation route={route} />)

    fireEvent.press(getByTestId('SaveRecord.Button'))
    expect(AsyncStorage.getItem).toBeCalledWith('brushingRecords');
})

test('Button fires without params', () => {
    const { getByTestId } = render(<BrushingRecord navigation={navigation} route={{ params: null }} />)

    fireEvent.press(getByTestId('SaveRecord.Button'))
    expect(AsyncStorage.getItem).toBeCalledWith('brushingRecords');
    expect(AsyncStorage.setItem).toBeCalledWith('brushingRecords', '[]');
})

test('header is Edit Record', () => {
    const { getByText } = render(<BrushingRecord navigation route={route} />)

    expect(getByText('Edit Record')).not.toBeNull()
})

test('header is Add Record', () => {
    const { getByText } = render(<BrushingRecord navigation route={{ params: null }} />)

    expect(getByText('Add Record')).not.toBeNull()
})

test('minutes value is equal to 2', () => {
    const { getByPlaceholderText, getAllByDisplayValue } = render(
        <BrushingRecord
            navigation
            route={{ params: null }}
        />
    )

    fireEvent.changeText(
        getByPlaceholderText('Minutes'),
        '2'
    );
    const minutes = getAllByDisplayValue('2');
    expect(minutes).toHaveLength(1)
})

test('seconds value is equal to 34', () => {
    const { getByPlaceholderText, getAllByDisplayValue } = render(
        <BrushingRecord
            navigation
            route={{ params: null }}
        />
    )

    fireEvent.changeText(
        getByPlaceholderText('Seconds'),
        '34'
    );
    const seconds = getAllByDisplayValue('34');
    expect(seconds).toHaveLength(1)
})