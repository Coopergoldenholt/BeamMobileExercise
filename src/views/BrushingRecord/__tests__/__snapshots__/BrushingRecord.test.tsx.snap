// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "padding": 15,
      "paddingTop": 50,
    }
  }
>
  <Text
    style={
      Object {
        "alignSelf": "center",
        "fontSize": 22,
        "fontWeight": "bold",
        "paddingBottom": 25,
      }
    }
    testID="Header.Text"
  >
    Edit Record
  </Text>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "white",
        "borderColor": "grey",
        "borderRadius": 5,
        "borderWidth": 1,
        "height": 50,
        "justifyContent": "center",
        "opacity": 1,
        "paddingLeft": 5,
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 20,
        }
      }
    >
      November 9, 2021 2:36 AM
    </Text>
  </View>
  <Text
    style={
      Object {
        "fontSize": 18,
        "paddingBottom": 10,
        "paddingTop": 20,
      }
    }
  >
    Duration:
  </Text>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "center",
        }
      }
    >
      <TextInput
        keyboardType="number-pad"
        onChangeText={[Function]}
        placeholder="Minutes"
        style={
          Object {
            "backgroundColor": "white",
            "borderColor": "grey",
            "borderRadius": 5,
            "borderWidth": 2,
            "height": 30,
            "marginRight": 5,
            "minWidth": 75,
            "paddingLeft": 5,
          }
        }
        value="2"
      />
      <Text>
        Minutes
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "center",
        }
      }
    >
      <TextInput
        keyboardType="number-pad"
        onChangeText={[Function]}
        placeholder="Seconds"
        style={
          Object {
            "backgroundColor": "white",
            "borderColor": "grey",
            "borderRadius": 5,
            "borderWidth": 2,
            "height": 30,
            "marginRight": 5,
            "minWidth": 75,
            "paddingLeft": 5,
          }
        }
        value="34"
      />
      <Text>
        Seconds
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
        "paddingTop": 20,
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "alignSelf": "center",
          "backgroundColor": "tomato",
          "borderRadius": 5,
          "height": 40,
          "justifyContent": "center",
          "opacity": 1,
          "width": 150,
        }
      }
    >
      <Text>
        Cancel
      </Text>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "alignSelf": "center",
          "backgroundColor": "#20a9ff",
          "borderRadius": 5,
          "height": 40,
          "justifyContent": "center",
          "opacity": 1,
          "width": 150,
        }
      }
      testID="SaveRecord.Button"
    >
      <Text
        style={
          Object {
            "color": "white",
          }
        }
      >
        Save
      </Text>
    </View>
  </View>
</View>
`;
